### PR TITLE
add untracked_files kwarg to porcelain.status

### PR DESCRIPTION
**Pull request by dtrifiro**
porcelain.status: add untracked_files kwarg

Status calls can be quite slow when large untracked directories are present because they are being walked.
Using `untracked_files="normal"` returns untracked files and directories (behaving like `git status -unormal`,
which is git's default), using `untracked_files="all"` (dulwich default) also returns untracked files in
untracked directories, `untracked_files="no"` returns no untracked files (like `git status -uno`).

Note that old behaviour is preserved by having `untracked_files` default to "all".

### CODE REVIEW

File: dulwich/tests/test_porcelain.py [hunk 1]
```diff
@@ -2167,7 +2178,32 @@ def test_get_untracked_paths_subdir(self):
                     self.repo.open_index(),
                     exclude_ignored=True,
                 )
-            )
+            ),
+        )
+
+    def test_get_untracked_paths_walk(self):
+        os.mkdir(os.path.join(self.repo.path, "dir"))
+        os.mkdir(os.path.join(self.repo.path, "dir/subdir"))
```
Comment (efiop) [inline dulwich/tests/test_porcelain.py:2186]:
```suggestion
        os.mkdir(os.path.join(self.repo.path, "dir", "subdir"))
```

Comment (efiop) [reply dulwich/tests/test_porcelain.py:2186]: Seems like there is inconsitency in this test. Sometimes use use separate parts and sometimes you hardcode `/`.

File: dulwich/porcelain.py [hunk 2]
```diff
@@ -1155,12 +1155,16 @@ def determine_wants(remote_refs, **kwargs):
             _import_remote_refs(r.refs, remote_name, fetch_result.refs)
 
 
-def status(repo=".", ignored=False):
+def status(repo=".", ignored=False, walk_untracked=True):
```
Comment (efiop) [inline dulwich/porcelain.py:1158]: Should it be `untracked_files` to match the cli flag?

File: dulwich/tests/test_porcelain.py [hunk 3]
```diff
@@ -1842,13 +1842,18 @@ def test_status_all(self):
         mod_path = os.path.join(self.repo.path, "bar")
         add_path = os.path.join(self.repo.path, "baz")
         us_path = os.path.join(self.repo.path, "blye")
-        ut_path = os.path.join(self.repo.path, "blyat")
+        ut_path = os.path.join(self.repo.path, "untracked_file")
+        os.mkdir(os.path.join(self.repo.path, "untracked_dir"))
+        ut_subfile_path = os.path.join(self.repo.path, "untracked_dir/file")
```
Comment (efiop) [inline dulwich/tests/test_porcelain.py:1847]:
```suggestion
        ut_subfile_path = os.path.join(self.repo.path, "untracked_dir", "file")
```
hence the failing windows test

Comment (dtrifiro) [reply dulwich/tests/test_porcelain.py]: Fixed everywhere

File: dulwich/porcelain.py [hunk 4]
```diff
@@ -1215,14 +1232,21 @@ def _walk_working_dir_paths(frompath, basepath, prune_dirnames=None):
...
+    walk=True,
+):
     """Get untracked paths.
 
     Args:
       frompath: Path to walk
       basepath: Path to compare to
       index: Index to check against
       exclude_ignored: Whether to exclude ignored paths
+      walk: Whether to walk untracked paths.
```
Comment (efiop) [inline dulwich/porcelain.py:1249]:
```suggestion
      walk: Whether to walk untracked paths
```

File: dulwich/porcelain.py [hunk 5]
```diff
@@ -1251,15 +1285,13 @@ def prune_dirnames(dirpath, dirnames):
         if not is_dir:
             ip = path_to_tree_path(basepath, ap)
             if ip not in index:
-                if (
-                    not exclude_ignored
-                    or not ignore_manager.is_ignored(
-                        os.path.relpath(ap, basepath)
-                    )
+                if not exclude_ignored or not ignore_manager.is_ignored(
+                    os.path.relpath(ap, basepath)
```
Comment (efiop) [inline dulwich/porcelain.py:1289]: Unrelated change?

File: dulwich/porcelain.py [hunk 6]
```diff
@@ -1243,6 +1268,15 @@ def prune_dirnames(dirpath, dirnames):
                         os.path.join(os.path.relpath(path, frompath), "")
                     )
                 del dirnames[i]
+            elif not walk:
```
Comment (efiop) [inline dulwich/porcelain.py:1271]: `walk` seems like a very generic term that doesn't really describe what is going on. Not that `normal` vs `all` are very descriptive themselves, but maybe let's just pass the policy down here and decide what to do in one place instead of introducing another translation layer with `walk`?

File: dulwich/tests/test_porcelain.py [hunk 7]
```diff
@@ -2167,7 +2184,32 @@ def test_get_untracked_paths_subdir(self):
...
+    def test_get_untracked_paths_walk(self):
+        os.mkdir(os.path.join(self.repo.path, "dir"))
+        os.mkdir(os.path.join(self.repo.path, "dir", "subdir"))
+        with open(os.path.join(self.repo.path, "dir", "file"), "w") as f:
+            f.write("foo")
+        with open(os.path.join(self.repo.path, "dir", "subdir", "subfile"), "w") as f:
+            f.write("foo")
+
+        self.assertEqual(
+            {"dir/"},
```
Comment (pmrowla) [inline dulwich/tests/test_porcelain.py:2199]: need to account for windows paths here with `os.path.join("dir", "")`

File: dulwich/porcelain.py [hunk 8]
```diff
@@ -1155,12 +1155,20 @@ def determine_wants(remote_refs, **kwargs):
...
       repo: Path to repository or repository object
       ignored: Whether to include ignored files in untracked
+      untracked_files: How to handle untracked files:
+            - "no": Show no untracked directories
+                  (like `git status -uno`).
+            - "all": show all files in untracked directories
+                  (like `git status -uall`, dulwich default).
+            - "normal": shows untracked files and directories.
+                  (like `git status -unormal`, git default).
+
```
Comment (pmrowla) [inline dulwich/porcelain.py:1171]: More of a question for @jelmer, but I'm not sure whether we actually need to support `all` vs `normal` here. It could just be `untracked_files=True` instead, where true=all and false=no. 

(The same way that existing `ignored` option just supports true/false rather than the full set of git's `traditional/no/matching` modes for ignored files)

Comment (jelmer) [reply dulwich/porcelain.py:1171]: Ideally we'd match the git command-line here. 

For convenience, we could support both strings and booleans as value for untracked_files?

**Comment by jelmer** [comment]
Any news on this; anything I can do to help unblock?

**Comment by dtrifiro** [comment]
Hi Jelmer,
I ended up noticing that this patch would yield slightly different results compared to `git status`, and I ended up deciding dropping it to avoid having mismatching behaviour between dulwich and git.

Just for the record, for a dummy repo:
```console
$ tree
├── tracked_with_ignored
│   ├── tracked_file
│   └── untracked
│       └── ignored
├── untracked_empty
│   ├── ignored
│   └── untracked_empty
│       └── untracked_empty
├── untracked_empty_dir
└── untracked_nonempty_dir
    └── file

7 directories, 4 files

$ git status --porcelain                                                                                                                                                                                  
?? .gitignore
?? untracked_nonempty_dir/
$ python -c 'from dulwich.porcelain import status; _, _, u = status(walk_untracked=False); [print(el) for el in u]'                                                                                           
untracked_empty/
untracked_nonempty_dir/
untracked_empty_dir/
tracked_with_ignored/untracked/
```
It seemed too much of an effort to replicate's git's [exact behaviour](https://github.com/git/git/blob/277cf0bc36094f6dc4297d8c9cef79df045b735d/dir.c#L1851) with dubious advantages.

Closing this and opening https://github.com/jelmer/dulwich/pull/964.
